app/api/v1/endpoints/control/agents.py

```
@@ -0,0 +1,368 @@
+) -> OffsetPaginatedResponse[AgentOut]:
+    """
+    List agents with offset-based pagination and filtering.
+
+    Access is scoped to agents in projects the user has access to.
+    """
+    try:
+        # Get user's accessible projects
+        accessible_projects = _get_accessible_projects(current_user)
+
+        if not accessible_projects:
+            raise ProjectAccessDeniedError(detail="User has no project access")
```

Ruff (0.15.0)

[warning] Abstract `raise` to an inner function

(TRY301)